File: backend/app/schemas/tax_operations.py
from __future__ import annotations

from datetime import date, datetime
from decimal import Decimal
from typing import Literal
import uuid

from pydantic import Field, model_validator

from app.schemas import BaseSchema


SourceModule = Literal["restaurant_pos", "retail_pos", "takeaway_pos", "purchasing", "manual"]
TaxDirection = Literal["output", "input"]
TaxCategory = Literal["standard", "zero", "exempt"]
ExportType = Literal["vat_sales", "vat_purchases", "pp30_summary", "wht_pnd3", "wht_pnd53", "etax_manifest", "tax_archive"]


class TaxLedgerIngest(BaseSchema):
    source_module: SourceModule
    tax_direction: TaxDirection
    tax_category: TaxCategory = "standard"
    source_document_type: str = Field(min_length=1, max_length=40)
    source_document_id: str = Field(min_length=1, max_length=100)
    line_key: str = Field(default="summary", min_length=1, max_length=100)
    source_status: str | None = Field(default=None, max_length=30)
    document_number: str = Field(min_length=1, max_length=100)
    document_date: date
    branch_id: uuid.UUID | None = None
    brand_id: uuid.UUID | None = None
    counterparty_name: str | None = Field(default=None, max_length=255)
    counterparty_tax_id: str | None = Field(default=None, max_length=20)
    counterparty_branch_code: str | None = Field(default=None, max_length=10)
    base_amount: Decimal = Field(ge=0)
    tax_amount: Decimal = Field(ge=0)
    total_amount: Decimal = Field(ge=0)
    vat_rate: Decimal = Field(ge=0, le=100)
    status: Literal["posted", "reversed", "excluded"] = "posted"
    source_event_id: str | None = Field(default=None, max_length=100)
    source_event_at: datetime | None = None

    @model_validator(mode="after")
    def validate_tax_values(self) -> "TaxLedgerIngest":
        if self.tax_category in {"zero", "exempt"} and self.tax_amount != 0:
            raise ValueError("รายการอัตรา 0% หรือยกเว้น ต้องไม่มีภาษี")
        if abs((self.base_amount + self.tax_amount) - self.total_amount) > Decimal("0.02"):
            raise ValueError("ยอดฐานภาษีและภาษีรวมไม่ตรงกับยอดรวม")
        return self


class TaxPeriodAction(BaseSchema):
    year: int = Field(ge=2000, le=2200)
    month: int = Field(ge=1, le=12)
    branch_id: uuid.UUID | None = None
    reason: str | None = Field(default=None, max_length=500)


class TaxIssueResolve(BaseSchema):
    status: Literal["resolved", "ignored"]
    note: str = Field(min_length=3, max_length=500)


class TaxExportCreate(BaseSchema):
    year: int = Field(ge=2000, le=2200)
    month: int = Field(ge=1, le=12)
    branch_id: uuid.UUID | None = None
    export_type: ExportType
    reason: str | None = Field(default=None, max_length=500)

File: frontend/src/types/taxOperations.ts
export type TaxSummary = {
  output_base: string;
  output_tax: string;
  input_base: string;
  input_tax: string;
  net_tax: string;
  wht_amount: string;
};

export type TaxLedgerRow = {
  id: string;
  source_module: string;
  tax_direction: "output" | "input";
  tax_category: string;
  document_number: string;
  document_date: string;
  counterparty_name: string | null;
  counterparty_tax_id: string | null;
  base_amount: string;
  tax_amount: string;
  total_amount: string;
  vat_rate: string;
  status: string;
  reconciliation_status: string;
};

export type TaxIssue = {
  id: string;
  issue_code: string;
  severity: "warning" | "error" | "blocker";
  status: "open" | "resolved" | "ignored";
  message: string;
  source_document_type: string | null;
  source_document_id: string | null;
  resolution_note: string | null;
};

export type TaxWhtRow = {
  id: string;
  certificate_number: string;
  issue_date: string;
  supplier_name: string;
  supplier_tax_id: string | null;
  tax_entity_type: string;
  wht_type: string;
  base_amount: string;
  wht_amount: string;
};

export type TaxExportRow = {
  id: string;
  export_type: string;
  status: string;
  filename: string;
  row_count: number;
  base_amount: string;
  tax_amount: string;
  content_sha256: string;
  generated_at: string;
};

export type TaxOperationsDashboard = {
  year: number;
  month: number;
  branch_id: string | null;
  summary: TaxSummary;
  period: { id: string | null; status: string; ledger_sha256: string | null; reviewed_at?: string | null; closed_at?: string | null; reopened_at?: string | null; reopen_reason?: string | null };
  ledger: TaxLedgerRow[];
  issues: TaxIssue[];
  wht: TaxWhtRow[];
  exports: TaxExportRow[];
  readiness: { ready_to_close: boolean; open_blockers: number; open_warnings: number; pending_reconciliation: number; configured: boolean };
};

export type TaxExportType = "vat_sales" | "vat_purchases" | "pp30_summary" | "wht_pnd3" | "wht_pnd53" | "etax_manifest" | "tax_archive";
